### PR TITLE
[SC-190] CFN custom resource for tagging EC2 instances

### PR DESCRIPTION
**Pull request by zaro0508**
The idea behind this cloudformation custom resource is to get information
from a synapse user profile request[1] and apply UserProfile[2] info as
tags on cloud ressources (buckets, EC2s, etc..).  This is a cleaner
alternative to using the custom scripts in Sage-Bionetworks/service-catalog-utils
to set EC2 tags.

Some benefits:
* Easier develop and maintain
* Consistency with S3 bucket tagger
* De-coupled from the SC CFN templates so it can be applied to any EC2
  resource.

[1] https://rest-docs.synapse.org/rest/GET/userProfile/profileId.html
[2] https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/UserProfile.html

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,6 +34,21 @@ the only new tag added is an `OwnerEmail` tag, whose value looks like
 [Synapse](https://www.synapse.org/) user name. Synapse provides email addresses
 for all user names.
 
+### EC2 Instance
+
+Create a custom resource in your cloud formation template. Here's an example:
```
Comment (zaro0508) [inline README.md:39]: cloud formation/cloudformation

File: set_instance_tags/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+crhelper
```
Comment (zaro0508) [inline set_instance_tags/requirements.txt:1]: scope these to specific versions

File: template.yaml [hunk 3]
```diff
@@ -1,21 +1,21 @@
 AWSTemplateFormatVersion: '2010-09-09'
 Transform: AWS::Serverless-2016-10-31
-Description: Custom Resource to sets tags on a S3 bucket
+Description: Custom Resource to sets tags on a EC2 instances and S3 buckets
 
 Globals:
   Function:
-    Timeout: 180
+    Timeout: 100
```
Comment (zaro0508) [inline template.yaml:7]: should be 180

File: template.yaml [hunk 4]
```diff
@@ -44,6 +44,45 @@ Resources:
               - 's3:PutBucketTagging'
             Resource: '*'
 
+  SetInstanceTagsFunction:
+    Type: AWS::Serverless::Function
+    Properties:
+      CodeUri: set_instance_tags/
+      Handler: app.handler
+      Role: !GetAtt SetInstanceTagsFunctionRole.Arn
+      Runtime: python3.6
```
Comment (zaro0508) [inline template.yaml:53]: should by python3.7

File: Pipfile [hunk 5]
```diff
@@ -13,8 +13,8 @@ pre-commit = "*"
 aws-sam-cli = "*"
 boto3 = "*"
 botocore = "*"
-requests = "*"
-crhelper = "*"
+requests = "~=2.23.0"
+crhelper = "~=2.0.6"
```
Comment (tthyer) [inline Pipfile:18]: +1 on adding version numbers -- don't want to do same for aws packages too?

**Review by tthyer — CHANGES_REQUESTED** [review]
Looks good! I wonder if you should name it resource-tagger rather than synapse-tagger though? If there were other tagging operations you want to move here or future ones?

**Comment by ahayden** [comment]
This also means an EC2 does not need rights to modify tags on itself or on other EC2s, so we can reduce the policy permissions on that.

**Comment by zaro0508** [comment]
> Looks good! I wonder if you should name it resource-tagger rather than synapse-tagger though? If there were other tagging operations you want to move here or future ones?

good suggestion @tthyer.  i was just looking for some feedback on this idea before doing more work.  seems like you guys think it's the better approach so I will continue with it.

**Comment by zaro0508** [comment]
i'm not sure i like the name change idea now @tthyer.  The functionality in this lambda is not a generic tagger it is very specific to taking synapse info and applying that info as tags to resources.  I think i will just keep the name as-ias

File: set_instance_tags/app.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+      principal_arn_value = tag.get('Value')
+      principal_id = principal_arn_value.split('/')[-1]
+      return principal_id
+  else:
+    raise ValueError('Could not derive a provisioningPrincipalArn from tags')
+
+
+def get_synapse_email(synapse_id):
+  '''Use the synapse id to look up the synapse user name'''
+  synapse_url = f'https://repo-prod.prod.sagebase.org/repo/v1/userProfile/{synapse_id}'
```
Comment (brucehoff) [inline set_instance_tags/app.py:67]: you should use the Synapse python client here.  See https://python-docs.synapse.org/build/html/Client.html#synapseclient.Synapse.getUserProfile

File: set_instance_tags/app.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  user_profile = response.json()
+  log.debug(f'Synapse user profile response: {user_profile}')
+  user_name = user_profile.get('userName')
+  synapse_email = f'{user_name}@synapse.org'
+  return synapse_email
+
+
+def get_owner_email(principal_id):
+  '''Derive an email from a id'''
+  email_check = re.compile(r"[^@]+@[^@]+\.[^@]+")
```
Comment (brucehoff) [inline set_instance_tags/app.py:79]: from _an_ id

It's not clear to me why a 'principal_id' would turn out to be an email address.   Where does such an email address come from?

**Review by brucehoff — CHANGES_REQUESTED** [review]

File: Pipfile (hunk 5 above)
Comment (zaro0508) [reply Pipfile:16]: fixed

File: .travis.yml [hunk 8]
```diff
@@ -19,15 +19,14 @@ jobs:
...
-        - pre-commit run --all-files
-        - sam build
-        - python -m pytest tests/ -v
+        - pipenv run pre-commit run --all-files
+        - pipenv run sam build
+        - pipenv run python -m pytest tests/ -v
     - stage: deploy
       script:
-        - sam build
+        - pipenv run sam build
```
[the remark this answers is not in the document]
Comment (zaro0508) [reply .travis.yml:27]: add pipenv in front runs in the pipenv virtual environment.  this virtual env is a clean environment containing only dependencies from the Pipfile

File: set_instance_tags/app.py (hunk 6 above)
Comment (zaro0508) [reply set_instance_tags/app.py:67]: is this a requirement?  i'm borrowing code,  `set_bucket_tags` function that @tthyer wrote, that already works.  if you feel strongly about this then I would like to implement it with a follow on PR to this change.

File: set_instance_tags/app.py (hunk 7 above)
Comment (zaro0508) [reply set_instance_tags/app.py:79]: the principal id is the synapse id which is used to retrieve the username and return  `<synapse user name>@synapse.org` as the email address.  isn't that the synapse email?

Comment (tthyer) [reply set_instance_tags/app.py:79]: When the original code was written, we had not completed implementing Synapse login, and were still using real IAM user accounts and manually setting OwnerEmail at least part of the time. At that time our principal ids mostly actually were email addresses. @zaro0508, up to you if you want to rewrite this function to assume that the principal ID will always be a Synapse ID now.

Comment (brucehoff) [reply set_instance_tags/app.py:79]: The principal_id is a number, not an email address.  For example, yours is 3377358, @zaro0508 (see https://www.synapse.org/#!Profile:3377358)

File: set_instance_tags/app.py (hunk 6 above)
Comment (brucehoff) [reply set_instance_tags/app.py:67]: This code (lines 67-70) duplicates code in the Synapse Python client so the best practice would be do just use the single command from that package.  We have a working example of using the client here: https://github.com/Sage-Bionetworks-IT/lambda-budgets.  I do not have a strong opinion whether the change is added to this PR or added in a follow-up PR.

File: set_instance_tags/app.py (hunk 7 above)
Comment (tthyer) [reply set_instance_tags/app.py:79]: @brucehoff You're correct, that is the case _now_. It was not always the case. I wrote a comment above to provide some historical context for everyone.

Comment (zaro0508) [reply set_instance_tags/app.py:79]: will remove email_check on a follow on PR.

**Pull request merged**